[Backend] Feature(auth): Implement authentication endpoints (Login and Registration with JWT) and tests #7

Trocador opened the pull request:
🎯Purpose
- Implement authentication endpoints (/register, /login, /me) with validations and JWT middleware.
- Document all changes and ensure proper functionality through manual and automated testing.

🔧Changes Implemented
Routes (auth-routes.ts)
- POST /api/auth/register
- Registers a new user with email and password validation.
- Error handling: duplicate email → 409 Conflict, password < 8 characters → 400 Validation error.
- POST /api/auth/login
- Logs in with valid credentials.
- Generates JWT with payload { id, email }.
- Error handling: invalid credentials → 401 Invalid credentials.
- GET /api/auth/me
- Protected route returning authenticated user data.
- Uses AuthMiddleware.validateToken to validate JWT.
- Error handling: invalid/expired token → 401 Unauthorized, user not found → 404 User not found.
Middleware (AuthMiddleware.validateToken)
- Validates Bearer <token> format.
- Decodes JWT using HS256 algorithm.
- Assigns req.userId for controllers.
- Handles invalid token, expired token, and unexpected errors.
Controller (ExpressAuthController.getMe)
- Retrieves req.userId from middleware.
- Calls getMeUseCase to fetch user data.
- Error handling: 401 Unauthorized, 404 User not found, 500 Internal server error.

Typing
- Defined JWTPayload:
```
export interface JWTPayload {
  id: string;
  email: string;
  iat?: number;
  exp?: number;
}
```
- Extended Express Request to include userId.

You can check the details of documentation and changes in:
https://docs.google.com/document/d/1QUez-P_qFi8J8EVUQxpkQshNWBblS-vqZL2KkJBCqTM/edit?usp=sharing

🧪Tests Performed
✅ Success Cases
- User registration → 201 Created.
- Login with valid credentials → 200 OK with access_token.
- /me with valid token → 200 OK with user data.

🚨 Error Cases
- Invalid or tampered token → 401 Invalid token.
- Expired token → 401 Token expired.
- Missing token in header → 401 Unauthorized.
- Duplicate email registration → 409 Conflict.
- Password shorter than 8 characters → 400 Validation error.
- Login with non-existent email or wrong password → 401 Invalid credentials.
- Deleted user with valid token → 404 User not found.

To check the tests of the endpoints and another tests performed:
https://docs.google.com/document/d/1Mn5-8id5kct34OCi0CNHWq_HPtZLyhWGEe-fd-VBF8Q/edit?usp=sharing


⚙️ Test Execution
- Manual (Postman): verified all scenarios with and without tokens(verify the before url)
- Manual runner (src/tests/index.ts):
```
import { testAuthSignup } from './application/auth/AuthSignup.test';
import { testAuthLogin } from './application/auth/AuthLogin.test';

async function runAllTests() {
  await testAuthSignup();
  await testAuthLogin();
}

runAllTests();
```
- Executed with:
npx ts-node src/tests/index.ts
- Future option: migrate to Jest + Supertest for CI/CD integration.

And finally the test execution
with comand:
`npx ts-node src/tests/index.ts`

we have the result:
<img width="1471" height="483" alt="image" src="https://github.com/user-attachments/assets/18228686-39b0-4a90-87fe-1ab6226291dd" />

📚 Validation Checklist
- [x] Registration and login working.
- [x] JWT generated and validated correctly.
- [x] Middleware handles token errors.
- [x] /me returns correct user data.
- [x] Error cases return proper HTTP codes.
- [x] Manual runner executes basic tests.

Trocador committed:
feat: files modified to check the requeriments of the ticket

Trocador committed:
feat: addded tests unitary to signup and login

Ditmar reviewed (changes requested):
The PR adds username uniqueness checks, repository methods, controller response shape changes, test scaffolding and type augmentation. Main wins: added domain VO for username, repository interface update and Prisma + mock implementations, and basic tests. Critical issues: several import paths are inconsistent (absolute vs relative) and will break at runtime; uniqueness is enforced only at application level (race condition) and needs DB unique constraints; username/email comparisons lack normalization (case-insensitive checks); controller removed a structured log and changed response shape without clear contract (may break clients) and possibly mismatches with use-case return shape; tests are ad-hoc console scripts (no assertions/test runner) and use hardcoded salt; domain validation throws generic Error types and contains a typo. Recommend addressing import path bugs first (will cause failures), adding DB-level uniqueness and normalization, and improving tests and error handling.

### ℹ️ Comentarios Adicionales

Se identificaron **10 comentarios adicionales** en líneas que no están en el diff actual. Estos comentarios se incluyen aquí como sugerencias generales:

**1. src/auth/application/auth-signup/auth-signup.ts:2**
The new import uses an absolute path 'auth/domain/auth-user-name' which is inconsistent with the rest of the file (other domain imports use relative paths) and will likely fail at runtime or in certain build setups.
*Sugerencia:* Use the same relative import style used elsewhere, e.g.:
import { AuthUserName } from '../../domain/auth-user-name';
(or update path aliases/config so absolute path works across the project)

**2. src/auth/application/auth-signup/auth-signup.ts:33**
The code checks username uniqueness at application level by calling repository.findByName. This prevents obvious duplicates, but it's prone to a race condition (two concurrent requests can pass this check and both create the same username). Also comparisons are done raw — no normalization (case, whitespace) is applied which can allow logically-duplicate usernames.
*Sugerencia:* Keep this application-level check but enforce uniqueness at the DB level (unique index on username). Normalize before checking/storing (e.g. trim + lowercase) or ensure AuthUserName VO enforces normalization. Example:
const normalized = name.trim().toLowerCase();
const existingUserByName = await this.repository.findByName(new AuthUserName(normalized));
// and ensure a UNIQUE constraint on the users.name column in the DB migration

**3. src/auth/domain/auth-user-password.ts:3**
Minimum password length was increased from 6 to 8, which is fine, but the code throws a generic Error and the inline comment contains a typo. Using generic Error makes it harder for callers to distinguish validation errors vs runtime errors.
*Sugerencia:* Use a domain-specific ValidationError (or at least a dedicated Error subclass) and centralize messages. Fix comment grammar. Example:
if (value.length < 8) {
  throw new ValidationError('Password must be at least 8 characters long');
}
(Also consider adding optional complexity requirements or delegating to a password policy module.)

**4. src/auth/domain/repository/auth-repository.ts:1**
You've added findByName to the repository interface which is appropriate for the new application behavior. Two important considerations: 1) the repository returns domain AuthUser which may include the password — ensure callers only get sensitive fields when required; 2) the interface change requires all adapters to implement the new method (mock/prisma updated here but double-check other adapters).
*Sugerencia:* Document that name should be unique and normalized (e.g. lowercase). Consider naming the method findByUsername for clarity (if 'name' could be ambiguous). Also consider returning a DTO or limiting fields when not necessary (e.g. exclude password unless caller requests it).

**5. src/auth/infrastructure/controller/express-auth-controller.ts:92**
The controller now returns { access_token, token_type } instead of forwarding the use-case result. Ensure this response shape is intentional and consistent with API contract (OpenAPI/docs). Also the earlier logger.info call was removed — auditing log entries for successful logins can be valuable, but be careful not to log PII (email) at info level.
*Sugerencia:* If the intended public contract is the OAuth-style response, include expires_in and scope if applicable. Example:
res.status(200).json({ access_token: result.token, token_type: 'Bearer', expires_in: result.expiresIn });
Reintroduce structured logging (but avoid logging email/stdout):
logger.info('user.login', { userId: result.userId, outcome: 'success' });
Ensure all clients are updated (or keep the legacy shape for backward compatibility).

**6. src/auth/infrastructure/repository/auth-mock-repository.ts:2**
Import path 'auth/domain/auth-user-name' is absolute and inconsistent with other imports in this file — likely to fail unless special TS path aliases are configured.
*Sugerencia:* Use the relative import to match the project style, e.g.:
import { AuthUserName } from '../../domain/auth-user-name';

**7. src/auth/infrastructure/repository/auth-mock-repository.ts:11**
findByName compares raw values (u.name.getValue() === name.getValue()). This is fine in tests but will behave differently from a DB that uses case-insensitive collation or if usernames are expected to be normalized (eg 'John' vs 'john').
*Sugerencia:* Normalize values for comparison (consistent with production behavior) e.g.:
this.users.find(u => u.name.getValue().toLowerCase() === name.getValue().toLowerCase());
And document that AuthUserName VO should already normalize/validate the input.

**8. src/auth/infrastructure/repository/auth-prisma-repository.ts:3**
New import for AuthUserName uses a relative path which is correct here, but double-check consistency across files (some other files used absolute before). Also, adding findByName to Prisma adapter is good but there are important considerations below.
*Sugerencia:* Ensure the import matches your project conventions. If you use path aliases, make it consistent in all files. Prefer '../../domain/auth-user-name' as used elsewhere.

**9. src/auth/infrastructure/repository/auth-prisma-repository.ts:34**
The Prisma queries use findUnique({ where: { email: ... } }) and similar for name. This assumes a UNIQUE constraint exists on those columns. Also the code fetches the full user record; in some cases you should restrict selected fields (e.g. avoid fetching password when not needed). There is no DB error handling around Prisma calls (e.g. transient DB errors).
*Sugerencia:* Make sure DB migrations create UNIQUE indexes for email and username. Use select to fetch only required fields and wrap DB calls to translate infrastructure errors into repository-level errors. Example:
const user = await this.prisma.user.findUnique({ where: { email: email.getValue() }, select: { id: true, name: true, email: true, password: true, createdAt: true } });
// wrap in try/catch and rethrow a repository error or let a higher-level error mapper handle it

**10. src/auth/infrastructure/routes/auth-routes.ts:22**
You changed the public route from '/signup' to '/register'. This is a breaking API change for any clients expecting /signup. Also comment indicates this was changed for a ticket but no versioning or redirection is present.
*Sugerencia:* If this is a breaking change, document it and bump API version (eg /v1/auth/register) or support both routes for a deprecation period:
authRoutes.post('/signup', (req, res) => authController.signup(req, res));
authRoutes.post('/register', (req, res) => authController.signup(req, res));
Add notes to OpenAPI/docs and update clients.


---

**Conclusión:** Do not merge yet. Fix the import path errors (they will break runtime), add DB-level unique constraints for username/email and normalize values before comparison, improve error handling (use domain-specific validation errors), and convert the ad-hoc console tests into proper unit tests with assertions. Also confirm the API response shape change and logging adjustments are intentional and documented (or provide backward compatibility). These issues are critical (imports, uniqueness/race conditions, tests) and should be resolved before approval.

Trocador commented on src/tests/application/auth/AuthLogin.test.ts:
```diff
@@ -0,0 +1,54 @@
+import { AuthLogin } from '../../../auth/application/auth-login/auth-login';
```
I resolve installing Jest:
package.json

```
"scripts": {
   "test": "jest --runInBand",
.
.
.
}
"devDependencies": {
"@types/jest": "^30.0.0",
...
"jest": "^30.2.0",
...
"supertest": "^7.1.4",
"ts-jest": "^29.4.6",
.
.
.
}
```

and modified the files AuthSignup.test.ts and AuthLogin.test.ts
You can see the changes inthe respective files.
To a correct execution of this, we need configure the jest.configure file:

```
const { createDefaultPreset } = require("ts-jest");

const tsJestTransformCfg = createDefaultPreset().transform;

/** @type {import("jest").Config} **/
module.exports = {
  preset: 'ts-jest',
  testEnvironment: "node",
  moduleNameMapper: {
    '^@config$': '<rootDir>/src/config/config.ts',
    '^@logger$': '<rootDir>/src/logger/index.ts',
    '^@prisma$': '<rootDir>/src/generated/client/index.ts',
  },
  transformIgnorePatterns: [
    '/node_modules/(?!uuid)',
  ],
  transform: {
    ...tsJestTransformCfg,
  },
};
```

and added a new file in src\__mocks__\uuid.ts:

`export const v4 = () => 'mocked-uuid';`

Trocador commented on src/tests/application/auth/AuthSignup.test.ts:
```diff
@@ -0,0 +1,54 @@
+import { AuthSignup } from '../../../auth/application/auth-signup/auth-signup';
```
Same resolved, info in the previous comment

Trocador commented on src/types/express/index.d.ts:
```diff
@@ -0,0 +1,5 @@
+declare module 'express-serve-static-core' {
```
i resolved this to replace with src/types/express/index.d.ts:

```
declare global {
  namespace Express {
    interface Request {
      userId?: string;
    }
  }
}
```
I put this file for an error of typed with express, it's necessary

Trocador commented:
I try to include the suggestions with the next addeds and eraseds
Added of imports relatives like:

`import { AuthUserName } from '../../domain/auth-user-name';`

Normalization with src\auth\domain\auth-user-name.ts:

```
export class AuthUserName {
  private readonly value: string;

  constructor(name: string) {
    const normalized = name.trim().toLowerCase();

    if (!normalized || normalized === '') {
      throw new Error('User name cannot be empty');
    }
    if (normalized.length < 3) {
      throw new Error('User name must be at least 3 characters long');
    }

    this.value = normalized;
  }
```
and added in schema.prisma:

```
model User {
  id           String        @id @default(uuid())
  email        String        @unique
  name         String        @unique
  password     String
  createdAt    DateTime      @default(now()) @map("created_at")
```
for a unique name and email

Corrected the generic errors with a specific class in src\auth\domain\auth-user-password.ts:

```
import { ValidationError } from './errors/validation-error';
export class AuthUserPassword {
  constructor(private readonly value: string) {
    if (!value || value.trim() === '') {
      throw new ValidationError('Password cannot be empty');
    }
    if (value.length < 8) {
      throw new ValidationError('Password must be at least 8 characters long'); //added requeriments from ticket 8 characters
    }
  }
```
Using a existing class src\auth\domain\errors\validation-error.ts:

```
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}
```

Corrected controller response shape y logging with reestructured in:

```
async signup(req: Request, res: Response): Promise<void> {
.
.
. 
logger.info(
        {
          userId: user.id.getValue(),
          email: user.email.getValue(),
          outcome: 'success',
        },
        'user.signup'
      );
.
.
.
async login(req: Request, res: Response): Promise<void> {
.
.
.
logger.info({ userId: result.user.id, outcome: 'success' }, 'user.login');
```
   
Corrected findbyName implemented on all adapters, for example on src\auth\infrastructure\repository\auth-mock-repository.ts:

```
async findByName(name: AuthUserName): Promise<AuthUser | null> {
    const user = this.users.find(
      u => u.name.getValue() === name.getValue().toLowerCase()
    );
    return user ?? null;
  }
```

and on  src\auth\infrastructure\repository\auth-prisma-repository.ts:

```
async findByName(name: AuthUserName): Promise<AuthUser | null> {
    const user = await this.prisma.user.findUnique({
      where: { name: name.getValue() },
    });
    if (!user) return null;

    return new AuthUser(
      user.id,
      user.name,
      user.email,
      user.password,
      user.createdAt
    );
  }
```
Migration of tests to use Jest using describe, it and expect. More details on the previous comments resolveds.
the config of Jest is on the file .eslintrc.js:

```
env: {
        node: true,
        es6: true,
        jest: true
    },

```
To fix an errors with Jest

The content of the file config of jest.config.js:

```
const { createDefaultPreset } = require("ts-jest");

const tsJestTransformCfg = createDefaultPreset().transform;

/** @type {import("jest").Config} **/
module.exports = {
  preset: 'ts-jest',
  testEnvironment: "node",
  moduleNameMapper: {
    '^@config$': '<rootDir>/src/config/config.ts',
    '^@logger$': '<rootDir>/src/logger/index.ts',
    '^@prisma$': '<rootDir>/src/generated/client/index.ts',
  },
  transformIgnorePatterns: [
    '/node_modules/(?!uuid)',
  ],
  transform: {
    ...tsJestTransformCfg,
  },
};
```

Corrections of breaks on clients for routes, added and suported src\auth\infrastructure\routes\auth-routes.ts:

```
authRoutes.post('/signup', (req, res) => authController.signup(req, res));
authRoutes.post('/register', (req, res) => authController.signup(req, res));
```

And finally a view of the new tests:

<img width="945" height="273" alt="image" src="https://github.com/user-attachments/assets/daa0442b-76fc-4444-a194-51d40d150c44" />


I hope these changes are sufficient to close this PR

Trocador committed:
fix: PR review requested changes applied

Ditmar reviewed (changes requested):
The implementation of authentication endpoints is mostly compliant with the ticket requirements. The code demonstrates good practices such as using bcrypt for password hashing and implementing unique username validation. However, there are some missing requirements, such as handling token expiration and validation in the login process. Additionally, the code could benefit from improved error handling and logging practices. Overall, the changes are a step in the right direction but need refinement for completeness.

### 🎫 Ticket Fulfillment (80/100)

**✅ Implemented Requirements:**
- Implementar el endpoint POST /auth/register para registrar un nuevo usuario.
- Implementar el endpoint POST /auth/login para autenticar a un usuario y devolver un token JWT.
- Validar que el correo no esté registrado previamente al registrar un usuario.
- Validar que la contraseña tenga un mínimo de 8 caracteres al registrar un usuario.
- Validar que el nombre de usuario sea único al registrar un usuario.
- Validar que el correo exista y la contraseña sea correcta al iniciar sesión.
- Usar bcrypt para hashear las contraseñas.
- Firmar el JWT con una clave secreta definida en las variables de entorno.
- Incluir pruebas unitarias para los servicios de registro y autenticación.

**❌ Missing Requirements:**
- Configurar la expiración del token JWT.
- Se valida correctamente la expiración y firma del token.


### ⚠️ Detected Side Effects

- The addition of a new route for registration may cause confusion if the existing signup route is not deprecated or documented properly.


### ℹ️ Additional Comments

There are **2 additional comments** on lines that are not in the current diff. These comments are included here as general suggestions:

**1. src/auth/application/auth-signup/auth-signup.ts:33**
The validation for unique usernames is implemented, which is good. However, ensure that the error handling is consistent across the application.
*Suggestion:* Consider using a centralized error handling mechanism to manage user-related errors.

**2. src/auth/infrastructure/controller/express-auth-controller.ts:99**
The response for the login endpoint includes the token and its expiration, which is good. However, ensure that the token expiration is validated before processing requests.
*Suggestion:* Implement middleware to check token validity and expiration before accessing protected routes.


---

**Conclusion:** While the implementation is on the right track, it requires adjustments to fully meet the ticket requirements, particularly regarding token expiration handling and consistent error management.

Ditmar commented on src/tests/application/auth/AuthLogin.test.ts:
```diff
@@ -0,0 +1,42 @@
+import { AuthLogin } from '../../../auth/application/auth-login/auth-login';
```
Tests are still not using a proper testing framework with assertions, which is essential for reliable testing.

**Suggestion:**
```typescript
Convert these tests to use Jest or a similar framework with proper assertions.
```

Ditmar commented on src/tests/application/auth/AuthSignup.test.ts:
```diff
@@ -0,0 +1,45 @@
+import { AuthSignup } from '../../../auth/application/auth-signup/auth-signup';
```
Similar to the login tests, these tests lack proper structure and assertions.

**Suggestion:**
```typescript
Implement these as formal unit tests with assertions using a testing framework.
```

Ditmar reviewed (changes requested):
The author has addressed several previous comments, particularly regarding username uniqueness checks and error handling. However, critical issues related to import paths and database-level uniqueness constraints remain unresolved. Additionally, the implementation of token expiration handling is still missing. New tests have been added, but they do not fully comply with the required structure. Overall, while progress has been made, significant issues persist.

### 🎫 Ticket Fulfillment (80/100)

**✅ Implemented Requirements:**
- Implementar el endpoint POST /auth/register para registrar un nuevo usuario.
- Implementar el endpoint POST /auth/login para autenticar a un usuario y devolver un token JWT.
- Validar que el correo no esté registrado previamente al registrar un usuario.
- Validar que la contraseña tenga un mínimo de 8 caracteres al registrar un usuario.
- Validar que el nombre de usuario sea único al registrar un usuario.
- Validar que el correo exista y la contraseña sea correcta al iniciar sesión.
- Usar bcrypt para hashear las contraseñas.
- Firmar el JWT con una clave secreta definida en las variables de entorno.
- Incluir pruebas unitarias para los servicios de registro y autenticación.

**❌ Missing Requirements:**
- Configurar la expiración del token JWT.
- Se valida correctamente la expiración y firma del token.


### ⚠️ Detected Side Effects

- The addition of a new route for registration may cause confusion if the existing signup route is not deprecated or documented properly.


### ℹ️ Additional Comments

There are **2 additional comments** on lines that are not in the current diff. These comments are included here as general suggestions:

**1. src/auth/application/auth-signup/auth-signup.ts:33**
The uniqueness check for usernames is still only at the application level, which can lead to race conditions. A database-level unique constraint is necessary to ensure data integrity.
*Suggestion:* Implement a unique constraint on the username field in the database schema.

**2. src/auth/infrastructure/controller/express-auth-controller.ts:99**
Token expiration handling is still missing, which is critical for security and functionality.
*Suggestion:* Implement middleware to validate token expiration before processing requests.



### 🔄 Re-review: Previous Comments Status

**✅ Addressed (1):**
- `src/auth/domain/auth-user-password.ts:15` — The implementation now uses a domain-specific ValidationError, which is an improvement.

**❌ Still pending (4):**
- `src/auth/application/auth-signup/auth-signup.ts:33` — The application-level uniqueness check is still present without a corresponding database-level unique constraint, leaving the potential for race conditions.
  > _Original comment:_ The code checks username uniqueness at application level by calling repository.findByName. This prevents obvious duplicates, but it's prone to a race condition (two concurrent requests can pass this check and both create the same username). Also comparisons are done raw — no normalization (case, whitespace) is applied which can allow logically-duplicate usernames.
- `src/auth/infrastructure/controller/express-auth-controller.ts:99` — Token expiration handling is still not implemented, which is a critical requirement.
  > _Original comment:_ The response for the login endpoint includes the token and its expiration, which is good. However, ensure that the token expiration is validated before processing requests.
- `src/tests/application/auth/AuthLogin.test.ts:1` — While tests have been added, they still do not utilize a proper testing framework with assertions, which is necessary for maintainability.
  > _Original comment:_ This file adds manual test scripts that log results to console instead of using a testing framework with assertions.
- `src/tests/application/auth/AuthSignup.test.ts:1` — Similar to the previous comment, the tests lack proper structure and assertions.
  > _Original comment:_ Same as above: tests are console-driven and lack assertions/test framework usage.

---

**Conclusion:** While some improvements have been made, critical issues regarding token expiration handling and database-level uniqueness constraints remain unresolved, necessitating further changes before approval.
